slim/slimlock.1
.TH slimlock 1 "January 12, 2023" "version 0.9"
.SH NAME
\fBslimlock\fP - Unholy Screen Locker
\fB
.SH SYNOPSIS
.nf
.fam C
\fBslimlock\fP [\-v]
.fam T
.fi
.SH DESCRIPTION
The Frankenstein's monster of screen lockers. Grafting SLiM and slock together
leads to blood, tears, and locked screens.
.SH OPTIONS
.TP
.B
\fB-v\fP
display version information
.SH CONFIGURATION
Slimlock reads the same configuration files you use for SLiM. It looks in 
\fICFGDIR/slim.conf\fP and \fICFGDIR/slimlock.conf\fP, where \fICFGDIR\fP 
is defined when built - usually \fI/etc\fP.
The options that are read from slim.conf are 
hidecursor, current_theme, background_color and background_style, 
screenshot_cmd, and welcome_msg. See the SLiM docs for more information.

slimlock.conf contains the following settings:

.TP
.B dpms_standby_timeout
number of seconds of inactivity before the screen blanks.
.BI "Default: " 60
.TP
.B dpms_off_timeout
number of seconds of inactivity before the screen is turned off.
.BI "Default: " 600
.TP
.B wrong_passwd_timeout
delay in seconds after an incorrect password is entered.
.BI "Default: " 2
.TP
.B passwd_feedback_msg
message to display after a failed authentication attempt.
.BI "Default: " "Authentication failed"
.TP
.B passwd_feedback_capslock
message to display after a failed authentication attempt if the CapsLock is on.
.BI "Default: " "Authentication failed (CapsLock is on)"
.TP
.B show_username
1 to show username on themes with single input field; 0 to disable.
.BI "Default: " 1
.TP
.B show_welcome_msg
1 to show SLiM's welcome message; 0 to disable.
.BI "Default: " 0
.TP
.B tty_lock
1 to disallow virtual terminals switching; 0 to allow.
.BI "Default: " 1
.TP
.B bell
1 to enable the bell on authentication failure; 0 to disable.
.BI "Default: " 1
.SH BUGS
Where the theme defines a welcome message or password prompt, \fBslimlock\fP
displays them in the wrong place.
.PP
Using the same theme as \fBslim\fP assumes the theme is appropriate. Most of
the themes out there (including our defaults) put the word "Login" on the
panel, which isn't quite right for a screen lock.
.PP
To alleviate both of these, you can specify a different theme in 
\fICFGDIR/slimlock.conf\fP, which will override the setting in 
\fICFGDIR/slim.conf\fP, but you still need to find (or create) a suitable
lock theme.
.SH "SEE ALSO"
.BR slim (1)


slim/ChangeLog
1.4.0 - 2023.01.21
	* - BREAKING CONFIG CHANGE - the position of the passwd_feedback in slim
	  is now relative to the screen, not the panel. This is now consistent 
	  with slimlock and with the session and other messages.
	* The selected session is now preserved over failed logins
	* When testing themes with the -p option, the size of the window can now
	  be specified, and the theme is previewed correctly for that size.
	* On exit from theme test, the "test message" now reports user and session
	* Failed login feedback is now disabled by default or if position negative
	* Applied some Devuan patches
	  - Set the XDG_SESSION_CLASS. This fixes Gentoo bug #727544 - failure to
	    register sessions with elogind in some instances. The (incorrect) bug
	    note in the man page has therefore been removed again.
	  - Use $DISPLAY instead of hardcoding :0.0 in sessionstart/stop commands
	  - No longer always builds slimlock if using PAM - it must be explicitly
	    enabled
	  - Fixed formatting in slimlock man page
	  - A couple of typos fixed in strings
	* Updated the README and encoded in UTF-8
	* Corrected the handling of the -n / -nodaemon option so that it doesn't 
	  swallow the next option if given as -n
	* Bug fixes in slimlock
	  - The active entry is for password, so show the password prompt, not the
	    username one
	  - Don't react to F1, as it isn't appropriate and used to blank the 
	    screen
	  - Keep it all on one screen even when the WM hasn't set a "primary" for
	    RandR purposes
	* Fix ticket #4 - the config parser now works properly. Previously it got
	  confused by option names that were substrings of other option names
	* Themes with "background_style color" no longer need a redundant image
	  file to make them work.
	  - This needed a bit of a re-write of the image handling, which also
	    improves efficiency
	* New default theme - the old one is retained as "original"
	* Some general documentation improvements (man pages, comments in example
	  files)

1.3.9 - 2022.11.18
    * Changed the handling of the "auth failed" message in DM mode so that it 
      remains on display until a key is pressed.
    * Added a command line option to specify the configuration file
    * Allow the logfile to be specified as "stderr" to write to stderr
      Also now writes all log messages to stderr if they are printed before 
      the log file is open, including when using the -p option.
    * Added a configuration option to retain the user name if authentication
      fails (which is probably more helpful if you just mistype your password)
    * Applied a modified version of the "wait for the X server" patch from guix
    * No longer unnecessarily re-opens an already open X display interface. 
      Similarly removed a redundant "get root window" call
    * Deleted some unused member variables and methods. Various other internal
      clean-up
    * Reverted the install location definition for the configuration file, as 
      the CMAKE_INSTALL_SYSCONFDIR wasn't right
    * Added a "bug" note to the man page, for the reliance on sessreg, which 
      causes occasional misbehaviour for some people.

1.3.8 - 2022.03.01
	* Fixed some bugs listed on bugs.gentoo.org:
	    832303 - failed to build with clang, Invalid suffix on string literal
	    580458 - open fd on /var/log/slim.log passed to session
	* Fixed handling of log stream so that all the code uses the same instance
	* Handle return values from calls, to clean up warn-unused-result warnings
	* Fixed "sessions" config option (the earlier patch was incomplete)
	* Several QA improvements
	* Updated cmake config - use standard install paths, don't force options

1.3.7 - 2022.01.30
	* Imported several bug fixes from the Gentoo package:
		libslim-cmake-fixes
		disable-ck-for-systemd
		reinstate sessions config option as an alternative to sessiondir
		systemd session compatibility
		remove zlib dependency
		envcpy-bad-pointer-arithmetic patch
		add-missing-libgen_h
		wrong comparison for XCreateGC error return (GCC11 needs it fixed!)
	* Fixed a bug in handling expired user accounts with PAM
	* Show a message on login failure (using the same method as slimlock)


 ---   The releases below were the original SLiM project   ---

1.3.6 - 2013.10.01
	* Merge slimlock.
	* Add support ^H (like backspace).
	* And fix some bugs.

1.3.5 - 2012.12.31
	* Support UTF8 string.
	* Add systemd service.
	* And fix some bugs.

1.3.4 - 2012.06.26
	* Replaced stderr writes function.
	* Fix numlock control.
	* Fix build with GLUT.
	* Fix PAM authentication.

1.3.3 - 2012.02.22
	* Change build system to CMake.
	* Add support ConsoleKit.
	* Fix some bugs....

1.3.2 - 2010.07.08
    * Add support xauth secret.
    * Add xnest_debug mode.

1.3.1 - 2008.09.26
    * Added focus_password config option for focusing password
      automatically when default_user is enabled
    * Added auto_login option
    * Fixed uninitialized daemonmode, see
      http://www.freebsd.org/cgi/query-pr.cgi?pr=114366
    * Fixed maximum length for password
    * Introduced customization options for session text:
      font, colors, position, shadows.

1.3.0 - 2006.07.14
    * Added PAM support by Martin Parm
    * Fixed segfault on exit when testing themes. Thanks
      to Darren Salt & Mike Massonnet
    * Fixed vt argument detection, thanks to Henrik Brix Andersen
    * Corrected reference to input_color in the default theme
    * Fixed default shell setting
    * Fix segfault when calling XCloseDisplay(NULL); thanks Uli Schlachter

1.2.6 - 2006.09.15
    * Bug #008167: Update pid when in daemon mode
    * Fixed warnings when compiling with -Wall. Thanks to
      KIMURA Masaru
    * Fixed major memory leaks with repeated login (bug #007535)

1.2.5 - 2006.07.24
    * hiding of the cursor is now an option (disabled
      by default) since some WMs does not re-initialize
      the root window cursor.
    * The X server is restarted when the user logs out.

slim/slimlock.cpp
/* slimlock
 * Copyright (c) 2010-2012 Joel Burget <joelburget@gmail.com>
 * Copyright (c) 2022-2023 Rob Pearce <slim@flitspace.org.uk>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 */

#include <cstdio>
#include <cstring>
#include <algorithm>
#include <sys/types.h>
#include <sys/ioctl.h>
#include <linux/vt.h>
#include <X11/keysym.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/extensions/dpms.h>
#include <security/pam_appl.h>
#include <pthread.h>
#include <err.h>
#include <signal.h>
#include <unistd.h>			// for usleep
#include <sys/stat.h>
#include <sys/file.h>
#include <errno.h>
#include <sys/file.h>
#include <fcntl.h>

#include "cfg.h"
#include "util.h"
#include "panel.h"
#include "const.h"

#undef APPNAME
#define APPNAME "slimlock"
#define SLIMLOCKCFG SYSCONFDIR"/slimlock.conf"

using namespace std;

bool AuthenticateUser();
static int ConvCallback(int num_msgs, const struct pam_message **msg,
						struct pam_response **resp, void *appdata_ptr);
void HandleSignal(int sig);
void *RaiseWindow(void *data);

// In the absence of a class instance to contain these, just make them file
// public, as they're needed by multiple functions
static Display* Dpy;
static int Scr;
static Window Root;
static Window win;
static Cfg* cfg;
static Panel* LoginPanel;

static pam_handle_t *pam_handle;

static const struct pam_conv conv = {ConvCallback, NULL};

static CARD16 dpms_standby, dpms_suspend, dpms_off, dpms_level;
static BOOL dpms_state, using_dpms;

static int term;	// Used by a C callback function


static void die(const char *errstr, ...)
{
	va_list ap;

	va_start(ap, errstr);
	vfprintf(stderr, errstr, ap);
	va_end(ap);
	exit(EXIT_FAILURE);
}

int main(int argc, char **argv)
{
	if((argc == 2) && !strcmp("-v", argv[1]))
		die ( APPNAME "-" VERSION ", © 2010-2012 Joel Burget\nUpdates © 2022-2023 Rob Pearce\n" );
	else if(argc != 1)
		die ( "usage: " APPNAME " [-v]\n" );

	void (*prev_fn)(int);

	// restore DPMS settings should slimlock be killed in the line of duty
	prev_fn = signal(SIGTERM, HandleSignal);
	if (prev_fn == SIG_IGN)
		signal(SIGTERM, SIG_IGN);

	// create a lock file to solve mutliple instances problem
	// /var/lock used to be the place to put this, now it's /run/lock
	// ...i think
	struct stat statbuf;
	int lock_file;

	// try /run/lock first, since i believe it's preferred
	if (!stat("/run/lock", &statbuf))
		lock_file = open ( "/run/lock/" APPNAME ".lock", O_CREAT | O_RDWR, 0666);
	else
		lock_file = open ( "/var/lock/" APPNAME ".lock", O_CREAT | O_RDWR, 0666);

	int rc = flock(lock_file, LOCK_EX | LOCK_NB);

	if(rc) {
		if(EWOULDBLOCK == errno)
			die(APPNAME" already running\n");
	}

	unsigned int cfg_passwd_timeout;

	/* Read configuration and theme */
	cfg = new Cfg;
	cfg->readConf(CFGFILE);
	cfg->readConf(SLIMLOCKCFG);
	string themeName = "";
	string themebase = "";
	string themefile = "";
	string themedir = "";

	themebase = string(THEMESDIR) + "/";
	themeName = cfg->getOption("current_theme");
	string::size_type pos;
	if ((pos = themeName.find(",")) != string::npos)
	{
		/* input is a set */
		themeName = cfg->findValidRandomTheme(themeName);
	}

	bool loaded = false;
	while (!loaded)
	{
		themedir =  themebase + themeName;
		themefile = themedir + THEMESFILE;
		if (!cfg->readConf(themefile))
		{
			if (themeName == "default")
			{
				cerr << APPNAME << ": Failed to open default theme file "
					 << themefile << endl;
				exit(ERR_EXIT);
			}
			else
			{
				cerr << APPNAME << ": Invalid theme in config: "
					 << themeName << endl;
				themeName = "default";
			}
		}
		else
		{
			loaded = true;
		}
	}

	const char *DisplayName = getenv("DISPLAY");
	if (!DisplayName)
		DisplayName = DISPLAY;

	Dpy = XOpenDisplay(DisplayName);
	if ( Dpy == 0 )
		die(APPNAME": cannot open display\n");

	/* Get screen and root window */
	Scr = DefaultScreen(Dpy);
	Root = RootWindow(Dpy, Scr);

	XSetWindowAttributes wa;
	wa.override_redirect = 1;
	wa.background_pixel = BlackPixel(Dpy, Scr);

	// Create a full screen window
	win = XCreateWindow(Dpy, Root,
			0, 0, DisplayWidth(Dpy, Scr), DisplayHeight(Dpy, Scr),
			0, DefaultDepth(Dpy, Scr), CopyFromParent,
			DefaultVisual(Dpy, Scr), CWOverrideRedirect | CWBackPixel,
			&wa);
	XMapWindow(Dpy, win);

	XFlush(Dpy);
	for (int len = 1000; len; len--) {
		if(XGrabKeyboard(Dpy, Root, True, GrabModeAsync, GrabModeAsync, CurrentTime)
			== GrabSuccess)
			break;
		usleep(1000);
	}
	XSelectInput(Dpy, win, ExposureMask | KeyPressMask);

	/* Create panel */
	LoginPanel = new Panel(Dpy, Scr, win, cfg, themedir, Panel::Mode_Lock);
	LoginPanel->HideCursor();

	int ret = pam_start(APPNAME, LoginPanel->GetName().c_str(), &conv, &pam_handle);
	// If we can't start PAM, just exit because slimlock won't work right
	if (ret != PAM_SUCCESS)
		die("PAM: %s\n", pam_strerror(pam_handle, ret));

	// disable tty switching
	if(cfg->getOption("tty_lock") == "1") {
		if ((term = open("/dev/console", O_RDWR)) == -1)
			perror("error opening console");

		if ((ioctl(term, VT_LOCKSWITCH)) == -1)
			perror("error locking console");
	}

	// Set up DPMS
	unsigned int cfg_dpms_standby, cfg_dpms_off;
	cfg_dpms_standby = Cfg::string2int(cfg->getOption("dpms_standby_timeout").c_str());
	cfg_dpms_off = Cfg::string2int(cfg->getOption("dpms_off_timeout").c_str());
	using_dpms = DPMSCapable(Dpy) && (cfg_dpms_standby > 0);
	if (using_dpms) {
		DPMSGetTimeouts(Dpy, &dpms_standby, &dpms_suspend, &dpms_off);

		DPMSSetTimeouts(Dpy, cfg_dpms_standby,
						cfg_dpms_standby, cfg_dpms_off);

		DPMSInfo(Dpy, &dpms_level, &dpms_state);
		if (!dpms_state)
			DPMSEnable(Dpy);
	}

	// Get password timeout
	cfg_passwd_timeout = Cfg::string2int(cfg->getOption("wrong_passwd_timeout").c_str());
	// Let's just make sure it has a sane value
	cfg_passwd_timeout = cfg_passwd_timeout > 60 ? 60 : cfg_passwd_timeout;

	pthread_t raise_thread;
	pthread_create(&raise_thread, NULL, RaiseWindow, NULL);

#if 0	// The DM code does this:
			/* Init Root */
			LoginPanel->setBackground();

			/* Show panel */
			LoginPanel->OpenPanel();
#endif

	// Main loop
	while (true)
	{
		LoginPanel->ResetPasswd();

		// AuthenticateUser returns true if authenticated
		if (AuthenticateUser())
			break;

		LoginPanel->WrongPassword(cfg_passwd_timeout);
	}

	// kill thread before destroying the window that it's supposed to be raising
	pthread_cancel(raise_thread);

	LoginPanel->ClosePanel();
	delete LoginPanel;

	// Get DPMS stuff back to normal
	if (using_dpms) {
		DPMSSetTimeouts(Dpy, dpms_standby, dpms_suspend, dpms_off);
		// turn off DPMS if it was off when we entered
		if (!dpms_state)
			DPMSDisable(Dpy);
	}

	XCloseDisplay(Dpy);

	close(lock_file);	// will inherently release the flock

	if(cfg->getOption("tty_lock") == "1") {
		if ((ioctl(term, VT_UNLOCKSWITCH)) == -1) {
			perror("error unlocking console");
		}
	}
	close(term);

	return 0;
}


static int ConvCallback(int num_msgs, const struct pam_message **msg,
						struct pam_response **resp, void *appdata_ptr)
{
	LoginPanel->EventHandler(Panel::Get_Passwd);

	// PAM expects an array of responses, one for each message
	if (num_msgs == 0 ||
		(*resp = (pam_response*) calloc(num_msgs, sizeof(struct pam_message))) == NULL)
		return PAM_BUF_ERR;

	for (int i = 0; i < num_msgs; i++) {
		if (msg[i]->msg_style != PAM_PROMPT_ECHO_OFF &&
			msg[i]->msg_style != PAM_PROMPT_ECHO_ON)
			continue;

		// return code is currently not used but should be set to zero
		resp[i]->resp_retcode = 0;
		if ((resp[i]->resp = strdup(LoginPanel->GetPasswd().c_str())) == NULL) {
			free(*resp);
			return PAM_BUF_ERR;
		}
	}

	return PAM_SUCCESS;
}

bool AuthenticateUser()
{
	return(pam_authenticate(pam_handle, 0) == PAM_SUCCESS);
}


void HandleSignal(int sig)
{
	// Get DPMS stuff back to normal
	if (using_dpms) {
		DPMSSetTimeouts(Dpy, dpms_standby, dpms_suspend, dpms_off);
		// turn off DPMS if it was off when we entered
		if (!dpms_state)
			DPMSDisable(Dpy);
	}

	if ((ioctl(term, VT_UNLOCKSWITCH)) == -1) {
		perror("error unlocking console");
	}
	close(term);

	LoginPanel->ClosePanel();
	delete LoginPanel;

	die(APPNAME": Caught signal; dying\n");
}

void* RaiseWindow(void *data)
{
	while(1) {
		XRaiseWindow(Dpy, win);
		sleep(1);
	}

	return (void *)0;
}
